packages/base/src/api/client.ts

```diff
@@ -57,31 +73,67 @@ export abstract class BaseApiClient {
     }
   }
 
-  // prettier-ignore
-  protected async apiCall<T>(fn: (api: AxiosInstance) => Promise<T>): Promise<T> {
-    const api = await this.init();
+  private async withRetry<TResponse>(
+    axiosInstance: AxiosInstance,
+    apiFunction: ApiFunction<TResponse>,
+    { retryCount, retryDelay }: { retryCount: number; retryDelay: number } = { retryCount: 0, retryDelay: 0 },
+  ): Promise<TResponse> {
     try {
-      return await fn(api);
+      await sleep(retryDelay);
+      return await apiFunction(axiosInstance);
     } catch (error: any) {
       // this means ID token has expired so we'll recreate session and try again
-      if (error.response && error.response.status === 401 && error.response.statusText === 'Unauthorized') {
+      if (isAuthenticationError(error)) {
         this.api = undefined;
 
         const api = await this.refresh();
-        return await fn(api);
+        return await this.withRetry(api, apiFunction, { retryCount, retryDelay });
       }
-      // Cloudflare error
-      if (error.response && error.response.status === 520 && error.response.data.includes('Cloudflare')) {
-        this.api = undefined;
 
-        const headersOverride: Record<string, string> = {
-          'Connection': 'upgrade',
-          'Upgrade': 'HTTP/2.0'
-        };
-        const api = await this.refresh({ headers: headersOverride });
-        return await fn(api);
+      const updatedRetryState = {
+        retryCount: retryCount + 1,
+        retryDelay: this.retryConfig.retryDelay(retryCount + 1, error),
+      };
+
+      if (updatedRetryState.retryCount > this.retryConfig.retries) throw error;
+
+      if (isCloudFlareError(error)) {
+        const apiWithUpgradeHeaders = await this.refresh({
+          headers: {
+            Connection: 'upgrade',
+            Upgrade: 'HTTP/2.0',
+          },
+        });
+
+        return await this.withRetry(apiWithUpgradeHeaders, apiFunction, updatedRetryState);
       }
+
+      if (await (this.retryConfig?.retryCondition?.(error) ?? true))
+        await this.withRetry(axiosInstance, apiFunction, updatedRetryState);
+
       throw error;
     }
   }
+
+  // prettier-ignore
+  protected async apiCall<TResponse>(apiFunction: ApiFunction<TResponse>): Promise<TResponse> {
+    const api = await this.init();
+    return this.withRetry(api, apiFunction)
+  }
 }
+
+const isAuthenticationError = (axiosError: AxiosError): boolean =>
+  axiosError.response?.status === 401 && axiosError.response?.statusText === 'Unauthorized';
+
+const isCloudFlareError = (axiosError: AxiosError): boolean =>
+  axiosError.response?.status === 520 && (axiosError.response?.data as string).includes('Cloudflare');
+
+export const exponentialDelay = (
+  retryNumber = 0,
+  _error: AxiosError | undefined = undefined,
+  delayFactor = 100,
+): number => {
+  const delay = 2 ** retryNumber * delayFactor;
+  const randomSum = delay * 0.2 * Math.random(); // 0-20% of the delay
+  return delay + randomSum;
+};
```

Comment:
I think the idea was to use [axios retry](https://www.npmjs.com/package/axios-retry) to abstract all this logic, is there a reason why you did the custom implementation?

Reply:
Yea, axios retry does not allow modifying the request (for cloudflare upgrade headers for example) also we qwould need to refactor much of how the api is managed if you follow the axios instance it get's recreated on refresh